interoperator/pkg/controller/sfplan/sfplan_controller.go

```diff
@@ -0,0 +1,176 @@
+limitations under the License.
+*/
+
+package sfplan
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	osbv1alpha1 "github.com/cloudfoundry-incubator/service-fabrik-broker/interoperator/pkg/apis/osb/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kubernetes "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Add creates a new SFPlan Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileSfPlan{Client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("sfplan-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to SFPlan
+	err = c.Watch(&source.Kind{Type: &osbv1alpha1.SFPlan{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileSfPlan{}
+
+// ReconcileSfPlan reconciles a SFPlan object
+type ReconcileSfPlan struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a SFPlan object and makes changes based on the state read
+// and what is in the SFPlan.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// +kubebuilder:rbac:groups=osb.servicefabrik.io,resources=sfplans,verbs=get;list;watch;create;update;patch;delete
+func (r *ReconcileSfPlan) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	// Fetch the SFPlan instance
+	instance := &osbv1alpha1.SFPlan{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	labels := instance.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	updateRequired := false
+	if serviceID, ok := labels["serviceId"]; !ok || instance.Spec.ServiceID != serviceID {
```

Comment:
All this controller is doing is updating the serviceId and planId labels from the spec. This is done so that we can list sfServices and sfPlans based on planId and serviceId. Ideally this should be done using a mutating webhook. But currently it is implemented as a controller. 
So to reconsile the cotoller reads from label and matches with spec, and if it does not match updates the label.

We plan to move it to a webhook.

Reply:
OK. Understood.